### PR TITLE
Fortran ABI debacle / MPI_SIZEOF

### PR DESCRIPTION
**Pull request by jsquyres**
Several fixes for the Fortran bindings on the v1.8 branch:
1. Ensure a proper shared library version number is applied to the libmpi_usempi_ignore_tkr.so library (it was always 0:0:0 before this).
2. Fix the ABI debacle with regards to MPI_SIZEOF (we screwed things up with regards to gfortran >=v4.9 in v1.8.2).  See the ompi/mpi/fortran/README-v1.8-ABI.txt file for details.

Note that the MPI_SIZEOF bindings were already fixed in the release branch in  514ea9f6cc59bcd6426ec3f524ffdfb4279491e0.

### CODE REVIEW

**Comment by rhc54** [comment]
@jsquyres has craig signed off on this?

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/84/
Test PASSed.

**Comment by jsquyres** [comment]
Not yet.  Working on getting a review.

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-release-v1.8-pr/85/
Test PASSed.

File: VERSION [hunk 1]
```diff
@@ -86,6 +86,7 @@ libmpi_so_version=7:0:6
 libmpi_cxx_so_version=2:3:1
 libmpi_mpifh_so_version=7:0:5
 libmpi_usempi_tkr_so_version=4:2:3
+libmpi_usempi_ignore_tkr_so_version=1:1:1
```
Comment (goodell) [inline VERSION]: Should probably be `1:0:1`.

File: config/ompi_setup_mpi_fortran.m4 [hunk 2]
```diff
@@ -645,6 +645,12 @@ end type test_mpi_handle],
                    [test $OMPI_BUILD_FORTRAN_USEMPI_BINDINGS -eq 1 -a \
                          $OMPI_FORTRAN_HAVE_IGNORE_TKR -eq 1])
 
+    # A dummy library is built for ABI reasons in v1.8.4 and beyond.
+    # See ompi/mpi/fortran/README-v1.8-ABI.txt for details.
+    AM_CONDITIONAL(OMPI_BUILD_FORTRAN_DUMMY_MPI_MODULE_LIBRARY,
+                   [test $OMPI_FORTRAN_HAVE_IGNORE_TKR -eq 1 && \
+                    test "$ompi_c_vendor" = "gnu"])
```
Comment (goodell) [inline config/ompi_setup_mpi_fortran.m4:662]: Please add a comment explaining C vs. Fortran compiler subtlety here (and why new code won't link against the dummy lib).

File: ompi/mpi/fortran/use-mpi-ignore-tkr/usempi-dummy.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+/*
+ * This is a dummy function, just to create a dummy library for ABI
+ * reasons in the v1.8 series.
+ *
+ * See ompi/mpi/fortran/README-v1.8-ABI.txt for details.
+ */
+int ompi_mpi_tkr_module_abi_symbol(void)
+{
+    return -1;
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/usempi-dummy.c]: Per review w/ Dave: pull in a symbol from libmpi_usempi_ignore_tkr.la to guarantee the dependency.

**Comment by goodell** [comment]
I reviewed w/ Jeff on the phone.  Jeff is going to fix a few minor things and re-test one case, but otherwise it's good to go.

File: VERSION (hunk 1 above)
Comment (jsquyres) [inline VERSION]: Fixed.

File: config/ompi_setup_mpi_fortran.m4 (hunk 2 above)
Comment (jsquyres) [inline config/ompi_setup_mpi_fortran.m4:662]: Fixed.

File: ompi/mpi/fortran/use-mpi-ignore-tkr/usempi-dummy.c (hunk 3 above)
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-ignore-tkr/usempi-dummy.c]: Fixed.

**Comment by jsquyres** [comment]
I address all of @goodell's point.

Ok Ralph, this is ready to go.

**Pull request merged**